search-provider/eks-discovery-feed-provider.c

```diff
@@ -359,11 +391,151 @@ handle_article_card_descriptions (EksDiscoveryFeedDatabaseContentProvider *skele
 
     return TRUE;
 }
-                       
+
+static void
+recent_news_articles_cb (GObject *source,
+                         GAsyncResult *result,
+                         gpointer user_data)
+{
+  EkncEngine *engine = EKNC_ENGINE (source);
+  DiscoveryFeedQueryState *state = user_data;
+
+  g_application_release (g_application_get_default ());
+
+  GError *error = NULL;
+  EkncDomain *domain = eknc_engine_get_domain_for_app (engine, state->provider->application_id, &error);
+  if (domain == NULL)
+    {
+      g_dbus_method_invocation_take_error (state->invocation, error);
+      discovery_feed_query_state_free (state);
+      return;
+    }
+
+  GSList *shards = eknc_domain_get_shards (domain);
+
+  GVariantBuilder shard_builder;
+  g_variant_builder_init (&shard_builder, G_VARIANT_TYPE_STRING_ARRAY);
+  for (GSList *l = shards; l; l = l->next)
+    {
+      g_autofree gchar *shard_path = NULL;
+      EosShardShardFile *shard = l->data;
+
+      g_object_get (shard, "path", &shard_path, NULL);
+      g_variant_builder_add (&shard_builder, "s", shard_path);
+    }
+
+  g_autoptr(EkncQueryResults) results = NULL;
+  if (!(results = eknc_engine_query_finish (engine, result, &error)))
+    {
+      g_dbus_method_invocation_return_gerror (state->invocation, error);
+      discovery_feed_query_state_free (state);
+      return;
+    }
```

Comment:
It should be - I've refactored this logic into a function shared by both handlers which does the check at the right place.